utilities/summarize_documentation/summarize_documentation.py

```diff
@@ -0,0 +1,112 @@
+# documentation is rebuilt.
+"""
+
+@dataclass(frozen=True)
+class NamespaceItem:
+    summary: str = ""
+    name: str = ""
+    namespace: str = ""
+    uid: str = ""
+    kind: str = ""
+
+@dataclass
+class Namespace:
+    summary: str = ""
+    uid: str = ""
+    name: str = ""
+    items = set()
+
+def items_of_kind(items, kind):
+    return [
+        {
+            "uid": item.uid,
+            "summary": item.summary
+        }
+        for item in
+        sorted(
+            (item for item in items if item.kind == kind),
+            key=(lambda item: item.uid)
+        )
+    ]
+
+@click.command()
+@click.argument("sources")
+@click.argument("output_path")
+def main(sources : str, output_path : str):
+    namespaces = defaultdict(Namespace)
+    output_path = Path(output_path)
+    for source in glob.glob(sources):
+        print(source)
+        with open(source, 'r', encoding='utf8') as f:
+            header, _ = frontmatter.parse(f.read())
+        if header['qsharp.kind'] == 'namespace':
+            namespaces[header['qsharp.name']].summary = header['qsharp.summary']
+            namespaces[header['qsharp.name']].name = header['qsharp.name']
+            namespaces[header['qsharp.name']].uid = header['uid']
+        else:
+            namespaces[header['qsharp.namespace']].items.add(NamespaceItem(
+                summary=header['qsharp.summary'],
+                name=header['qsharp.name'],
+                namespace=header['qsharp.namespace'],
+                uid=header["uid"],
+                kind=header["qsharp.kind"]
+            ))
+
+    for namespace_name, namespace in namespaces.items():
+        uid = namespace.uid or namespace_name
+        name = namespace.name or namespace_name
+        namespace_page = {
+            "uid": uid,
+            "name": name,
+            "summary": namespace.summary,
+            "operations": items_of_kind(namespace.items, "operation"),
+            "functions": items_of_kind(namespace.items, "function"),
+            "newtypes": items_of_kind(namespace.items, "udt")
+        }
+        
+        with open(output_path / f"{name.lower()}.yml", "w", encoding="utf8") as f:
+            f.write(namespace_comment + warning_comment + yaml.dump(namespace_page))
+
+    toc_page = [
+        {
+            "uid": namespace.name,
+            "name": namespace_name,
+            "items": [
+                {
+                    "name": item.name,
+                    "uid": item.uid
+                }
+                for item in sorted(
+                    namespace.items,
+                    key=lambda item: item.uid
+                )
+            ]
+        }
+        for namespace_name, namespace in sorted(
+            namespaces.items(),
+            key=lambda pair: pair[0]
+        )
+    ]
+    with open(output_path / "toc.yml", "w", encoding="utf8") as f:
+        f.write(warning_comment + yaml.dump(toc_page))
+
+if __name__ == "__main__":
+    main()
```

Comment:
What is the Python script used for? Instead of doc generator -> Markdown -> YAML, could the doc generator generate both Markdown and YAML directly?

Reply:
That's a bit tricky to do, as the YAML files have to summarize the Markdown files from multiple different assemblies (e.g.: to make a TOC for all Q# APIs in a single reference). I'll add a bit to it, but the purpose of this Python script is to produce those summary YAML files from a collection of different Markdown files. By contrast with our current approach, this makes documentation building stateless, as the documentation generation step no longer needs to read in an existing summary and update it with new content.

Reply:
My thought was to keep using the Markdown parsing parts of DocsParser, but to use the new rewrite steps to generate output instead of the YAML writing functionality in DocsParser. Happy to mark those methods with Obsolete if that would make sense?

Reply:
Yes, I think it would be good to mark these things as obsolete and keep just one setup also in the SDK (without the option to configure it to either be markdown or yaml). Then we can eventually remove these methods, giving us one project for the parsing of hte doc comments, one for the emission of the markdown, and a separate tool to convert the markdown to yaml. Does that sound reasonable?

Reply:
Sounds good, I'll remove the QsharpDocsGenerationMode property, then, and will lock everything down to the new Markdown mode. To support that, will mark any YAML output methods in DocsParser as Obsolete.